File: frontend/src/app/features/analysis/algorithms/thermal-asymmetry.ts
/**
 * Assimetria térmica entre as mãos, articulação por articulação, na **imagem estática**.
 *
 * Compara cada articulação com a sua correspondente do outro lado. É por articulação,
 * e não pela média da mão, porque inflamação articular é focal: uma MCP com 1,4 °C de
 * diferença, diluída entre dez articulações simétricas, vira 0,1 °C na média da mão —
 * um número que não diz nada.
 *
 * "Estática" é o termo do domínio, e não um eufemismo para "uma imagem só": ele opõe
 * esta leitura — as mãos em repouso, num instante — à termografia **dinâmica**, que é a
 * resposta ao estresse térmico ao longo do tempo. É por isso que o escopo está no nome
 * e não escondido na implementação: com uma sequência carregada, este algoritmo usa a
 * primeira captura **com articulação medida** e diz isso no relatório. A evolução do
 * reaquecimento é outra pergunta, e caberá a um algoritmo dinâmico.
 */

import {
  AlgorithmFrame,
  AlgorithmInput,
  AlgorithmJoint,
  AlgorithmResult,
  ResearchAlgorithm,
} from './algorithm.model';

/**
 * Cobertura de pele mínima para a medição entrar na comparação.
 *
 * Constante, e não parâmetro de tela: um botão exigiria descrever o parâmetro em
 * metadados e renderizar formulário, o que é máquina demais para uma demanda que não
 * existe. Está aqui, com nome, e muda-se em uma linha.
 */
const COBERTURA_MINIMA = 0.4;

interface Par {
  readonly label: string;
  /** E − D: positivo significa mão esquerda mais quente. */
  readonly assinada: number;
}

/**
 * Por que uma medição ficou de fora.
 *
 * Os dois motivos eram contados juntos e relatados como cobertura de pele, o que
 * descrevia errado o caso mais grave dos dois: uma ROI sem temperatura nenhuma não é
 * uma ROI mal coberta, é uma ROI que não mediu. Quem lê o relatório precisa saber qual
 * dos dois aconteceu, porque a ação é diferente.
 */
type MotivoDescarte = 'cobertura' | 'sem-temperatura';

interface Descartado {
  readonly label: string;
  readonly motivo: MotivoDescarte;
}

/** Por que esta medição não serve para comparar, ou `null` quando serve. */
function motivoDeDescarte(joint: AlgorithmJoint): MotivoDescarte | null {
  if (!Number.isFinite(joint.mean)) {
    return 'sem-temperatura';
  }
  return joint.skinCoverage < COBERTURA_MINIMA ? 'cobertura' : null;
}

/** Pareia as articulações da captura, separando o que não dá para comparar. */
function parear(joints: readonly AlgorithmJoint[]): {
  readonly pares: readonly Par[];
  readonly descartados: readonly Descartado[];
} {
  const esquerdas = new Map(
    joints.filter((j) => j.side === 'Esquerda').map((j) => [j.landmarkId, j]),
  );
  const direitas = new Map(
    joints.filter((j) => j.side === 'Direita').map((j) => [j.landmarkId, j]),
  );

  const pares: Par[] = [];
  const descartados: Descartado[] = [];

  for (const [landmarkId, esquerda] of esquerdas) {
    const direita = direitas.get(landmarkId);
    if (!direita) {
      continue; // não detectada do outro lado: não há par, e não é descarte
    }
    // O lado que falhou pior manda no motivo. Sem temperatura é falha de medição,
    // cobertura baixa é medição fraca, e relatar a segunda escondendo a primeira
    // mandaria o leitor conferir o enquadramento de uma ROI que nem mediu.
    const motivo = motivoDeDescarte(esquerda) ?? motivoDeDescarte(direita);
    if (motivo) {
      descartados.push({ label: esquerda.label, motivo });
      continue;
    }
    pares.push({
      label: esquerda.label,
      assinada: esquerda.mean - direita.mean,
    });
  }

  pares.sort((a, b) => Math.abs(b.assinada) - Math.abs(a.assinada));
  return { pares, descartados };
}

/** Qual mão está mais quente no par, ou `null` quando as duas medem igual. */
function ladoMaisQuente(assinada: number): 'esquerda' | 'direita' | null {
  if (assinada === 0) {
    return null;
  }
  return assinada > 0 ? 'esquerda' : 'direita';
}

/**
 * Os descartes em prosa, agrupados por motivo.
 *
 * Com um motivo só a frase não repete a contagem, porque "1 descartado: 1 por
 * cobertura" diz duas vezes a mesma coisa.
 */
function frasesDeDescarte(descartados: readonly Descartado[]): string {
  const rotulos: Record<MotivoDescarte, string> = {
    cobertura: `cobertura de pele abaixo de ${COBERTURA_MINIMA * 100}%`,
    'sem-temperatura': 'medição sem temperatura',
  };

  const grupos = (['cobertura', 'sem-temperatura'] as const)
    .map((motivo) => ({
      motivo,
      nomes: descartados.filter((d) => d.motivo === motivo).map((d) => d.label),
    }))
    .filter((grupo) => grupo.nomes.length > 0);

  const descrever = (grupo: (typeof grupos)[number]): string =>
    `${grupo.nomes.length} por ${rotulos[grupo.motivo]} (${grupo.nomes.join(', ')})`;

  if (grupos.length === 1) {
    const [grupo] = grupos;
    const palavra = grupo.nomes.length === 1 ? 'descartado' : 'descartados';
    return `${grupo.nomes.length} ${palavra} por ${rotulos[grupo.motivo]} (${grupo.nomes.join(', ')}).`;
  }
  return `${descartados.length} descartados, ${grupos.map(descrever).join(' e ')}.`;
}

/**
 * Aviso de escopo.
 *
 * O nome já diz "estática", mas quem lê só o relatório não vê o nome — e um texto sobre
 * 1 de 21 capturas passaria por um texto sobre as 21.
 */
function origem(frame: AlgorithmFrame, posicao: number, total: number): string | null {
  if (total === 1) {
    return null;
  }
  const estatica = 'é uma leitura estática, não a evolução do reaquecimento.';
  if (posicao > 0) {
    // Houve captura antes desta, e ela não tinha medição. Dizer "primeira captura"
    // aqui descreveria uma captura que não é esta — e esconderia o buraco, que é
    // justamente o que o frame vazio existe para mostrar.
    return (
      `Calculado sobre a captura de índice ${frame.captureIndex}, das ${total} carregadas: ` +
      `${posicao === 1 ? 'a anterior não tem' : `as ${posicao} anteriores não têm`} ` +
      `articulação medida. Com a ressalva, ${estatica}`
    );
  }
  // Nomeia a captura pelo que ela é, não por um instante: a basal fica fora do eixo
  // de reaquecimento (ver `rewarming-curve.ts`), e chamá-la de "instante inicial"
  // contradiria a curva e a linha do tempo, que já não lhe dão posição no tempo.
  const identificacao =
    frame.phase === 'baseline'
      ? 'a captura basal'
      : `a primeira captura, de índice ${frame.captureIndex},`;
  return `Calculado sobre ${identificacao} das ${total} carregadas: ${estatica}`;
}

export const thermalAsymmetry: ResearchAlgorithm = {
  slug: 'assimetria-termica-estatica',
  title: 'Assimetria térmica (imagem estática)',
  description:
    'Compara cada articulação com a correspondente do outro lado numa imagem estática, as mãos num instante, em repouso, e reporta as diferenças de temperatura da maior para a menor. Numa sequência, usa a primeira captura com medição.',

  run(input: AlgorithmInput): AlgorithmResult {
    // A primeira **com medição**, e não `frames[0]`: uma basal cujo alinhamento
    // falhou entra na sequência como frame de `joints: []`, e usá-la faria o
    // algoritmo relatar "nada detectado nas duas mãos" sobre uma sequência em que
    // 20 capturas foram medidas. O relatório diz qual captura usou.
    const posicao = input.frames.findIndex((f) => f.joints.length > 0);
    // Nenhuma tem medição: cai na primeira, e `parear` produz o insuficiente certo.
    const frame = posicao === -1 ? input.frames[0] : input.frames[posicao];

    // O painel já não chama `run` sem medição, mas a guarda não é por causa dele:
    // é para `run` ser total. A assinatura promete um `AlgorithmResult` para
    // qualquer `AlgorithmInput`, e um chamador futuro — um lote, um teste, outra
    // tela — não tem como saber de uma pré-condição que só existe no painel.
    if (!frame) {
      return {
        status: 'insufficient-data',
        summary: 'Não há captura analisada: não há medição sobre a qual calcular assimetria.',
        values: [],
      };
    }
    const { pares, descartados } = parear(frame.joints);

    if (pares.length === 0) {
      return {
        status: 'insufficient-data',
        summary:
          descartados.length > 0
            ? `Nenhum par pôde ser comparado. ${frasesDeDescarte(descartados)}`
            : 'Não há articulação detectada nas duas mãos: sem par correspondente, não há assimetria a calcular.',
        values: [],
      };
    }

    const maior = pares[0];
    const ladoDoMaior = ladoMaisQuente(maior.assinada);
    // Só contagens no texto, nenhum número com casa decimal, porque formatar é
    // trabalho da tela. Os valores vão em `values`, como números.
    const frases = [
      ladoDoMaior === null
        ? 'Nenhuma diferença de temperatura entre as mãos nos pares comparados.'
        : `Maior diferença na ${maior.label}, com a mão ${ladoDoMaior} mais quente.`,
      `${pares.length} ${pares.length === 1 ? 'par comparado' : 'pares comparados'}.` +
        (descartados.length > 0 ? ` ${frasesDeDescarte(descartados)}` : ''),
    ];

    const aviso = origem(frame, Math.max(posicao, 0), input.frames.length);
    if (aviso) {
      frases.push(aviso);
    }

    return {
      status: 'ok',
      summary: frases.join(' '),
      // O lado vai no rótulo, e não no sinal do número.
      //
      // `values` é magnitude, e a tabela da tela ordena e formata sem saber de
      // convenção de sinal. Antes só o maior par dizia qual mão estava mais quente,
      // na frase do resumo: as outras linhas mostravam "0,8 °C" e o leitor não tinha
      // como saber de que lado, apesar de o algoritmo já ter calculado isso.
      values: pares.map((par) => {
        const lado = ladoMaisQuente(par.assinada);
        return {
          label:
            lado === null ? `${par.label} (sem diferença)` : `${par.label} (${lado} mais quente)`,
          value: Math.abs(par.assinada),
          unit: '°C',
        };
      }),
    };
  },
};

File: backend/app/application/algorithms/thermal_asymmetry.py
"""Assimetria térmica entre as mãos, articulação por articulação, na imagem estática.

Compara cada articulação com a sua correspondente do outro lado. É por articulação, e
não pela média da mão, porque inflamação articular é focal: uma MCP com 1,4 °C de
diferença, diluída entre dez articulações simétricas, vira 0,1 °C na média da mão, um
número que não diz nada.

"Estática" é o termo do domínio, e não um eufemismo para "uma imagem só": ele opõe esta
leitura, as mãos em repouso num instante, à termografia dinâmica, que é a resposta ao
estresse térmico ao longo do tempo. Por isso o escopo está no nome e não escondido na
implementação: com uma sequência carregada, este algoritmo usa a primeira captura COM
articulação medida e diz isso no resumo. A evolução do reaquecimento é outra pergunta, e
caberá a um algoritmo dinâmico.

Veio do frontend, onde rodava em TypeScript sobre o estado da tela. O cálculo é o mesmo,
inclusive o corte de cobertura de pele: o que mudou foi de onde os números chegam.
"""

from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass
from typing import Literal

from app.domain.algorithms import (
    AlgorithmResult,
    AlgorithmStatus,
    AlgorithmValue,
    AnalysisCapture,
    JointMeasurement,
)

# Cobertura de pele mínima para a medição entrar na comparação.
#
# Constante, e não parâmetro de tela: um botão exigiria descrever o parâmetro em
# metadados e renderizar formulário, o que é máquina demais para uma demanda que não
# existe. Está aqui, com nome, e muda-se em uma linha.
COBERTURA_MINIMA = 0.4

MotivoDescarte = Literal["cobertura", "sem-temperatura"]


@dataclass(frozen=True, slots=True)
class _Par:
    label: str
    # E menos D: positivo significa mão esquerda mais quente.
    assinada: float


@dataclass(frozen=True, slots=True)
class _Descartado:
    label: str
    motivo: MotivoDescarte


def _motivo_de_descarte(medicao: JointMeasurement) -> MotivoDescarte | None:
    """Por que esta medição não serve para comparar, ou `None` quando serve.

    Os dois motivos já foram contados juntos e relatados como cobertura de pele, o que
    descrevia errado o mais grave dos dois: uma ROI sem temperatura nenhuma não é uma
    ROI mal coberta, é uma ROI que não mediu. Quem lê precisa saber qual dos dois
    aconteceu, porque a ação é diferente.
    """
    if medicao.temperature is None:
        return "sem-temperatura"
    return "cobertura" if medicao.skin_coverage < COBERTURA_MINIMA else None


def _sufixo(medicao: JointMeasurement) -> str:
    """`RIGHT_MCP_3` e `LEFT_MCP_3` viram `MCP_3`, que é o que pareia os dois lados.

    O prefixo sai a partir de `side`, que é coluna do catálogo, e não por corte de
    string no palpite: se a nomenclatura mudar, o catálogo é quem manda.
    """
    return medicao.joint_id.removeprefix(f"{medicao.side.upper()}_")


def _parear(
    medicoes: Sequence[JointMeasurement],
) -> tuple[list[_Par], list[_Descartado]]:
    """Pareia as articulações da captura, separando o que não dá para comparar."""
    esquerdas = {_sufixo(m): m for m in medicoes if m.side == "left"}
    direitas = {_sufixo(m): m for m in medicoes if m.side == "right"}

    pares: list[_Par] = []
    descartados: list[_Descartado] = []

    for sufixo, esquerda in esquerdas.items():
        direita = direitas.get(sufixo)
        if direita is None:
            # Não detectada do outro lado: não há par, e não é descarte.
            continue
        # O lado que falhou pior manda no motivo. Sem temperatura é falha de medição,
        # cobertura baixa é medição fraca, e relatar a segunda escondendo a primeira
        # mandaria o leitor conferir o enquadramento de uma ROI que nem mediu.
        motivo = _motivo_de_descarte(esquerda) or _motivo_de_descarte(direita)
        if motivo is not None:
            descartados.append(_Descartado(label=esquerda.label, motivo=motivo))
            continue
        # As duas temperaturas existem: `_motivo_de_descarte` já recusou o contrário.
        assert esquerda.temperature is not None and direita.temperature is not None
        pares.append(
            _Par(label=esquerda.label, assinada=esquerda.temperature - direita.temperature)
        )

    pares.sort(key=lambda par: abs(par.assinada), reverse=True)
    return pares, descartados


def _lado_mais_quente(assinada: float) -> str | None:
    """Qual mão está mais quente no par, ou `None` quando as duas medem igual."""
    if assinada == 0:
        return None
    return "esquerda" if assinada > 0 else "direita"


def _frases_de_descarte(descartados: Sequence[_Descartado]) -> str:
    """Os descartes em prosa, agrupados por motivo.

    Com um motivo só a frase não repete a contagem, porque "1 descartado: 1 por
    cobertura" diz duas vezes a mesma coisa.
    """
    rotulos: dict[MotivoDescarte, str] = {
        "cobertura": f"cobertura de pele abaixo de {COBERTURA_MINIMA * 100:.0f}%",
        "sem-temperatura": "medição sem temperatura",
    }

    grupos = [
        (motivo, [d.label for d in descartados if d.motivo == motivo])
        for motivo in ("cobertura", "sem-temperatura")
    ]
    grupos = [(motivo, nomes) for motivo, nomes in grupos if nomes]

    if len(grupos) == 1:
        motivo, nomes = grupos[0]
        palavra = "descartado" if len(nomes) == 1 else "descartados"
        return f"{len(nomes)} {palavra} por {rotulos[motivo]} ({', '.join(nomes)})."

    descricoes = " e ".join(
        f"{len(nomes)} por {rotulos[motivo]} ({', '.join(nomes)})" for motivo, nomes in grupos
    )
    return f"{len(descartados)} descartados, {descricoes}."


def _origem(captura: AnalysisCapture, posicao: int, total: int) -> str | None:
    """Aviso de escopo.

    O nome do algoritmo já diz "estática", mas quem lê só o resultado não vê o nome, e
    um texto sobre 1 de 21 capturas passaria por um texto sobre as 21.
    """
    if total == 1:
        return None

    estatica = "é uma leitura estática, não a evolução do reaquecimento."

    if posicao > 0:
        # Houve captura antes desta, e ela não tinha medição. Dizer "primeira captura"
        # aqui descreveria uma captura que não é esta, e esconderia o buraco, que é
        # justamente o que a captura sem medição existe para mostrar.
        faltantes = "a anterior não tem" if posicao == 1 else f"as {posicao} anteriores não têm"
        return (
            f"Calculado sobre a captura de índice {captura.capture_index}, "
            f"das {total} carregadas: {faltantes} articulação medida. "
            f"Com a ressalva, {estatica}"
        )

    # Nomeia a captura pelo que ela é, não por um instante: a basal fica fora do eixo de
    # reaquecimento, e chamá-la de "instante inicial" contradiria a curva e a linha do
    # tempo, que já não lhe dão posição no tempo.
    identificacao = (
        "a captura basal"
        if captura.is_baseline
        else f"a primeira captura, de índice {captura.capture_index},"
    )
    return f"Calculado sobre {identificacao} das {total} carregadas: {estatica}"


class ThermalAsymmetry:
    slug = "assimetria-termica-estatica"
    title = "Assimetria térmica (imagem estática)"
    description = (
        "Compara cada articulação com a correspondente do outro lado numa imagem "
        "estática, as mãos num instante, em repouso, e reporta as diferenças de "
        "temperatura da maior para a menor. Numa sequência, usa a primeira captura "
        "com medição."
    )

    def run(self, captures: Sequence[AnalysisCapture]) -> AlgorithmResult:
        # A primeira COM medição, e não a de índice 0: uma basal cujo alinhamento
        # falhou entra na sequência sem medição nenhuma, e usá-la faria o algoritmo
        # relatar "nada detectado nas duas mãos" sobre uma sequência em que 20 capturas
        # foram medidas. O resumo diz qual captura usou.
        posicao = next((i for i, c in enumerate(captures) if c.measurements), -1)
        # Nenhuma tem medição: cai na primeira, e `_parear` produz o insuficiente certo.
        captura = captures[max(posicao, 0)] if captures else None

        # A rota já não chama `run` sem análise, mas a guarda não é por causa dela: é
        # para `run` ser total. A assinatura promete um resultado para qualquer entrada,
        # e um chamador futuro não tem como saber de uma pré-condição que só existe na
        # borda.
        if captura is None:
            return AlgorithmResult(
                status=AlgorithmStatus.INSUFFICIENT_DATA,
                summary=(
                    "Não há captura analisada: não há medição sobre a qual calcular assimetria."
                ),
            )

        pares, descartados = _parear(captura.measurements)

        if not pares:
            return AlgorithmResult(
                status=AlgorithmStatus.INSUFFICIENT_DATA,
                summary=(
                    f"Nenhum par pôde ser comparado. {_frases_de_descarte(descartados)}"
                    if descartados
                    else "Não há articulação detectada nas duas mãos: sem par "
                    "correspondente, não há assimetria a calcular."
                ),
            )

        maior = pares[0]
        lado_do_maior = _lado_mais_quente(maior.assinada)

        # Só contagens no texto, nenhum número com casa decimal, porque formatar é
        # trabalho da tela. Os valores vão em `values`, como números.
        frases = [
            "Nenhuma diferença de temperatura entre as mãos nos pares comparados."
            if lado_do_maior is None
            else f"Maior diferença na {maior.label}, com a mão {lado_do_maior} mais quente.",
            f"{len(pares)} {'par comparado' if len(pares) == 1 else 'pares comparados'}."
            + (f" {_frases_de_descarte(descartados)}" if descartados else ""),
        ]

        aviso = _origem(captura, max(posicao, 0), len(captures))
        if aviso is not None:
            frases.append(aviso)

        return AlgorithmResult(
            status=AlgorithmStatus.OK,
            summary=" ".join(frases),
            # O lado vai no rótulo, e não no sinal do número: `values` é magnitude, e a
            # tabela da tela ordena e formata sem saber de convenção de sinal.
            values=[
                AlgorithmValue(
                    label=(
                        f"{par.label} (sem diferença)"
                        if (lado := _lado_mais_quente(par.assinada)) is None
                        else f"{par.label} ({lado} mais quente)"
                    ),
                    value=abs(par.assinada),
                    unit="°C",
                )
                for par in pares
            ],
        )


thermal_asymmetry = ThermalAsymmetry()
